rpc: de-flake TestHeartbeatHealthTransport

tbg opened the pull request:
The test uses a listener that is at some point closed. Closing the listener does
however not mean that all active connections become closed as well. This race
was already addressed throughout most of the test but not in the final
assertion. This is done in this commit.

Also catch some other unexpected errors that I saw while locally stress- testing
on OSX.

Fixes #18285.

Will cherry-pick this to 1.1 after merge.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18493)
<!-- Reviewable:end -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/rpc/context_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku59Xa483g5L9tYuC55:-Ku59Xa483g5L9tYuC56:b4acnjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L270)):*
> ```Go
> 	}{}
> 	ln := &interceptingListener{
> 		Listener: func() net.Listener {
> ```

I don't get this refactoring. Why the inline function? It's hard to read.

---

*[pkg/rpc/context_test.go, line 273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku57eNT4q-MAFl2w_Xk:-Ku57eNT4q-MAFl2w_Xl:b-5jbho4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L273)):*
> ```Go
> 			ln, err := net.Listen("tcp", util.TestAddr.String())
> 			if err != nil {
> 				t.Fatal(err)
> ```

I think this Fatal would be called on the wrong goroutine, no?

---

*[pkg/rpc/context_test.go, line 306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku57w8Du7gz3s4E5C3x:-Ku57w8Du7gz3s4E5C3y:b-umlkh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L306)):*
> ```Go
> 	})
> 
> 	closeConns := func() (int, error) {
> ```

pls comment what it returns

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-:-Ku59iMUqkMwdQJDY2qM:b-g2a5fb)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/rpc/context_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku59Xa483g5L9tYuC55:-Ku5iK1zYLYmaomKDg-4:bexjmsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L270)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't get this refactoring. Why the inline function? It's hard to read.
</blockquote></details>

The old code left an `err` in the outermost scope and twice I screwed up and checked something on that error instead of one I defined.

---

*[pkg/rpc/context_test.go, line 273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku57eNT4q-MAFl2w_Xk:-Ku5iR_VDtdTQtVOJZzg:bxujmum) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L273)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this Fatal would be called on the wrong goroutine, no?
</blockquote></details>

no, it's called on the main goroutine?

---

*[pkg/rpc/context_test.go, line 306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku57w8Du7gz3s4E5C3x:-Ku5ikie11h_QfW4cJKV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L306)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls comment what it returns
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-:-Ku5irMFl-voz4XMy8GO:b-814g8k)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/rpc/context_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku59Xa483g5L9tYuC55:-Ku5jATZt-xeBofNw3Dr:bwvpk6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L270)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The old code left an `err` in the outermost scope and twice I screwed up and checked something on that error instead of one I defined.
</blockquote></details>

then I'd give that err some other name. Or at least move the anonymous function to the top method scope.

---

*[pkg/rpc/context_test.go, line 273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku57eNT4q-MAFl2w_Xk:-Ku5j5_GNgTJqCgJgc7f:b-ouciru) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L273)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

no, it's called on the main goroutine?
</blockquote></details>

ah right, sorry. See - hard to read :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-:-Ku5jX9RKLy7Ufvo2Nh2:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented on pkg/rpc/context_test.go:
```diff
@@ -353,7 +368,15 @@ func TestHeartbeatHealthTransport(t *testing.T) {
 	// Should stay unhealthy despite reconnection attempts.
 	errUnhealthy := errors.New("connection is still unhealthy")
 	if err := util.RetryForDuration(100*clientCtx.heartbeatInterval, func() error {
```
It's somewhat unrelated to your fix, but this retry loop looks pretty silly. `RetryForDuration` keeps retrying until a nil error is returned to it or it times out. If it times out, it only returns the last error it received to its caller. Because this closure never returns a nil error, we're effectively calling it many times over the course of `100*clientCtx.heartbeatInterval`, but only ever examining the output of the last call.

tbg committed:
rpc: de-flake TestHeartbeatHealthTransport

The test uses a listener that is at some point closed. Closing the listener does
however not mean that all active connections become closed as well. This race
was already addressed throughout most of the test but not in the final
assertion. This is done in this commit.

Also catch some other unexpected errors that I saw while locally stress testing
on OSX.

Fixes #18285.

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/rpc/context_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku59Xa483g5L9tYuC55:-KuHifhpDNK-6oJrfQyS:bua4xed) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bdbbf875f2473998c8399773dfcac9b6e4e7140/pkg/rpc/context_test.go#L270)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

then I'd give that err some other name. Or at least move the anonymous function to the top method scope.
</blockquote></details>

Made the closure top-level, so I think it's more palatable now.

---

*[pkg/rpc/context_test.go, line 370 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493#-Ku0UQpTID4BvJeo9Z-O-r2-370:-KuHfzYz8ktUTezDY3NL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f52efdd5fc0e4f58f3230d06ea0782e5f14539c7/pkg/rpc/context_test.go#L370)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It's somewhat unrelated to your fix, but this retry loop looks pretty silly. `RetryForDuration` keeps retrying until a nil error is returned to it or it times out. If it times out, it only returns the last error it received to its caller. Because this closure never returns a nil error, we're effectively calling it many times over the course of `100*clientCtx.heartbeatInterval`, but only ever examining the output of the last call.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18493)*
<!-- Sent from Reviewable.io -->
